src/charm.py

```diff
@@ -9,264 +9,93 @@
-                Path(TEMPO_CA_FILE).write_text(ca_cert)
-                subprocess.run(["update-ca-certificates", "--fresh"])
-            return True
-
-        else:
-            self.tempo.clear_certs()
-
-            # clear from charm container too
-            Path(TEMPO_CA_FILE).unlink(missing_ok=True)
-            subprocess.run(["update-ca-certificates", "--fresh"])
-            return True
-
-    def pre_update_checklist(
-        self,
-    ) -> List[str]:
-        """Check if the charm as a whole is ready, and if not, return a list of failed checks."""
-        failures = []
-        failures.extend(self.tempo.pre_update_checklist())
-
-        if not self.tempo_cluster.get_tempo_config():
-            failures.append("tempo_cluster config not ready")
-
-        return failures
-
     def _on_collect_status(self, e: CollectStatusEvent):
-        if not self._container.can_connect():
-            e.add_status(WaitingStatus("Waiting for `tempo` container"))
-        if not self.model.get_relation("tempo-cluster"):
-            e.add_status(
-                BlockedStatus("Missing tempo-cluster relation to a tempo-coordinator charm")
-            )
-        elif not self.tempo_cluster.relation:
-            e.add_status(WaitingStatus("Tempo-Cluster relation not ready"))
-        if not self.tempo_cluster.get_tempo_config():
-            e.add_status(WaitingStatus("Waiting for coordinator to publish a tempo config"))
-
-        if role := self.tempo_role:
-            e.add_status(
-                ActiveStatus(
-                    "(all roles) ready." if role is TempoRole.all else f"{role.name} ready."
-                )
-            )
-        else:
-            logger.error(
-                f"`role` config value {self.config.get('role')!r} invalid: should "
-                f"be one of {list(TempoRole)}."
-            )
+        # add Tempo worker custom blocking conditions
+        if self.worker.roles and self.worker.roles[0] not in self._valid_roles:
             e.add_status(
                 BlockedStatus(f"Invalid `role` config value: {self.config.get('role')!r}.")
             )
 
     @property
     def tempo_endpoint(self) -> Optional[str]:
         """Tempo endpoint for charm tracing."""
-        return self.tempo_cluster.get_tracing_endpoint("otlp_http")
+        if endpoints := self.worker.cluster.get_tracing_receivers():
+            return endpoints.get("otlp_http", None)
 
     @property
     def ca_cert_path(self) -> Optional[str]:
         """CA certificate path for tls tracing."""
-        return TEMPO_CA_FILE if Path(TEMPO_CA_FILE).exists() else None
+        return CLIENT_CA_FILE
+
+    def pebble_layer(self, worker: Worker) -> Layer:
+        """Return a dictionary representing a Pebble layer.
+
+        Caller is responsible for checking whether the tempo role is valid before
+        calling this method.
+        """
+        role = "scalable-single-binary" if "all" in worker.roles else "".join(worker.roles)
```

Comment:
Erm, perhaps I'm confusing things, but worker.roles is supposed to be a list with only one item?

Why are we `"".join(["foo"])` instead of doing `worker.roles[0]`?

Also, perhaps, to give body to the 'caller's responsibility' disclaimer you put in the docstring, perhaps it would make sense to add a hard validation step here:

```python
roles = self.worker.roles
if not len(roles) == 1:
    raise RolesConfigurationError(...)
role = roles[0]
if not role in self._valid_roles:
    raise RolesconfigurationError(...)
```

Reply:
I'd rather fail hard here than push a nonsensical pebble layer.